packages/ses/src/memoize.js

```diff
@@ -0,0 +1,30 @@
+import { WeakMap, weakmapHas, weakmapGet, weakmapSet } from './commons.js';
+
+/**
+ * Given a one-argument function `fn` or a WeakMap-key compatible
+ * argument `arg`, returns `memoFn`, a memoizing form of that function
+ * that will only call `fn(arg)` for a given `arg` the first time.
+ * The `memoFn` function remembers the result and reuses it on subsequent
+ * calls with the same `arg`.
+ *
+ * See memoize.md for the Memoization Safety properties of `memoize`.
+ * (TODO turn into link once there's a URL)
+ *
+ * @template {any} A=[any]
```

Comment:
Simpler to have one param which is itself the function that takes one argument.

Reply:
That's what the code does. You mean the type param and function type?

Reply:
Hi @turadg , I tried to figure this out and could not. Could you provide an example? Thanks.

Reply:
First, I tried

```js
/**
 * @template {(arg: any) => any} F function type
 * @param {F} fn
 * @returns {F}
 */
export const memoize = fn => {
  const memo = new WeakMap();

  /** @type {F} */
  const memoFn = arg => {
```
but got 
```
src/memoize.js:19:9 - error TS2322: Type '(arg: any) => any' is not assignable to type 'F'.
  '(arg: any) => any' is assignable to the constraint of type 'F', but 'F' could be instantiated with a different subtype of constraint '(arg: any) => any'.

19   const memoFn = arg => {
           ~~~~~~


Found 1 error in src/memoize.js:19
```

Reply:
Then I tried
```js
/**
 * @template {(arg: any) => any} F function type
 * @param {F} fn
 * @returns {F}
 */
export const memoize = fn => {
  const memo = new WeakMap();

  /**
   * @param {Parameters<F>[0]} arg
   * @returns {ReturnType<F>}
   */
  const memoFn = arg => {
    ...
  return harden(memoFn);
};
```
and got
```
src/memoize.js:30:10 - error TS2322: Type '(arg: Parameters<F>[0]) => ReturnType<F>' is not assignable to type 'F'.
  '(arg: Parameters<F>[0]) => ReturnType<F>' is assignable to the constraint of type 'F', but 'F' could be instantiated with a different subtype of constraint '(arg: any) => any'.

30   return harden(memoFn);
            ~~~~~~~~~~~~~~

  src/memoize.js:30:10
    30   return harden(memoFn);
                ~~~~~~~~~~~~~~
    Did you mean to call this expression?


Found 1 error in src/memoize.js:30
```
IIUC the two error messages, they seem to have the same root cause. I often run into this problem, but I have previously had a compact capture to report. I would love to understand what to about this.

Reply:
cc @michaelfig

Reply:
To thwart these kind of "is assignable to the constraint of type 'F', but 'F' could be instantiated with..." errors, you need to do an explicit typecast.  The only syntax to do so in JSDoc is to use `/** @type {sometype} */ (expr)`.  The parentheses are *not* optional, and all together they form an expression with value `expr` and type `sometype`.

```js
  /**
   * @param {Parameters<F>[0]} arg
   * @returns {ReturnType<F>}
   */
  const memoFn = arg => {
    ...
  };
  harden(memoFn);
  // This cast beats the type system into submission so that it accepts memoFn really is type F. 
  return /** @type {F} */ (memoFn);
```

Reply:
Given that I would need a type case, would this change still be a net simplification?

Reply:
@turadg @michaelfig , after the above two attempts that IMO made things worse, how do you feel about proceeding with the current type declarations?